src/read.c

```diff
@@ -37,6 +37,8 @@
 
 struct ReaderState {
 
+ScannerState s;
```

Comment:
The ScannerState becomes part of the ReaderState (I've intended to do it this way for quite some time now - finally it happens, yay!)

One way to make this perhaps slightly nicer, and avoid some of the diffs in this PR permanently (i.e. not just with my crappy HACK commit) would be to switch `read.c` to C++, and then make `ReaderState` into a "subclass" of `ScannerState`. This way, a `ReaderState *` can be passed to any function expecting a `ScannerState`. 

But I have doubts whether that's enough to justify converting the whole file to C++...  ?!